src/dso_api/dynamic_api/filters/parser.py

```diff
@@ -0,0 +1,379 @@
+"""The engine for filtering from the query-string.
+
+This translates a query-string into ORM lookups using the Amsterdam Schema definitions.
+"""
+from __future__ import annotations
+
+import operator
+import re
+from datetime import datetime
+from functools import reduce
+from typing import Any
+
+from django.core.exceptions import ValidationError as DjangoValidationError
+from django.db import models
+from django.db.models import Q
+from django.utils.datastructures import MultiValueDict
+from gisserver.geometries import CRS
+from rest_framework.exceptions import PermissionDenied, ValidationError
+from schematools.contrib.django.models import DynamicModel
+from schematools.exceptions import SchemaObjectNotFound
+from schematools.permissions import UserScopes
+from schematools.types import DatasetFieldSchema, DatasetTableSchema
+from schematools.utils import to_snake_case
+
+from .values import str2bool, str2geo, str2isodate, str2number, str2time
+
+# Allow notations: field.subfield[lookup]
+RE_KEY = re.compile(r"\A(?P<path>[\w\d_-]+(?:\.[\w\d_-]+)*)(?:\[(?P<lookup>[a-z0-9A-Z_-]+)])?\Z")
```

Comment:
There's a `parse_filter` function in `dso_api.dynamic_api.permissions` to parse a filter into field name and lookup operator. It produces specific error messages and is (IMHO) more readable than this.

Reply:
I found out about that function later too. I'm not sure about the `permissions` module. We'll likely have to refactor that indeed. All filters and views already check for permissions now, so it looks like an additional "firewall" upfront to perform the same checks.

Reply:
chosen to leave as it is for now